bgpd/bgp_conditional_adv.c

```diff
@@ -0,0 +1,451 @@
+ */
+
+#include "bgpd/bgp_conditional_adv.h"
+
+const char *get_afi_safi_str(afi_t afi, safi_t safi, bool for_json);
+
+/* We just need bgp_dest node matches with filter prefix. So no need to
+ * traverse each path here.
+ */
+struct bgp_dest *bgp_dest_matches_filter_prefix(struct bgp_table *table,
+						struct filter *filter)
+{
+	uint32_t check_addr;
+	uint32_t check_mask;
+	struct in_addr mask;
+	struct bgp_dest *dest = NULL;
+	struct bgp_path_info *pi = NULL;
+	const struct prefix *dest_p = NULL;
+	struct filter_cisco *cfilter = NULL;
+	struct filter_zebra *zfilter = NULL;
+
+	if (filter->cisco) {
+		cfilter = &filter->u.cfilter;
+		for (dest = bgp_table_top(table); dest;
+		     dest = bgp_route_next(dest)) {
+			dest_p = (struct prefix *)bgp_dest_get_prefix(dest);
+			if (!dest_p)
+				continue;
+			pi = bgp_dest_get_bgp_path_info(dest);
+			if (!pi)
+				continue;
+			check_addr = dest_p->u.prefix4.s_addr
+				     & ~cfilter->addr_mask.s_addr;
+			if (memcmp(&check_addr, &cfilter->addr.s_addr,
+				   sizeof(check_addr))
+			    != 0)
+				continue;
+			if (cfilter->extended) {
+				masklen2ip(dest_p->prefixlen, &mask);
+				check_mask = mask.s_addr
+					     & ~cfilter->mask_mask.s_addr;
+				if (memcmp(&check_mask, &cfilter->mask.s_addr,
+					   sizeof(check_mask))
+				    != 0)
+					continue;
+			}
+			return dest;
+		}
+	} else {
+		zfilter = &filter->u.zfilter;
+		for (dest = bgp_table_top(table); dest;
+		     dest = bgp_route_next(dest)) {
+			dest_p = bgp_dest_get_prefix(dest);
+			if (!dest_p)
+				continue;
+			pi = bgp_dest_get_bgp_path_info(dest);
+			if (!pi)
+				continue;
+			if ((zfilter->prefix.family != dest_p->family)
+			    || (zfilter->exact
+				&& (zfilter->prefix.prefixlen
+				    != dest_p->prefixlen)))
+				continue;
+			else if (!prefix_match(&zfilter->prefix, dest_p))
+				continue;
+			else
+				return dest;
+		}
+	}
+	return NULL;
+}
+
+enum route_map_cmd_result_t
+bgp_check_rmap_prefixes_in_bgp_table(struct bgp_table *table,
+				     struct route_map *rmap)
+{
+	afi_t afi;
+	struct access_list *alist = NULL;
+	struct filter *alist_filter = NULL;
+	struct bgp_dest *dest = NULL;
+	struct route_map_rule *match = NULL;
+	enum route_map_cmd_result_t ret = RMAP_NOOP;
+
+	if (!is_rmap_valid(rmap))
+		return ret;
+
+	/* If several match commands are configured, all must succeed for a
+	 * given route in order for that route to match the clause (logical AND)
+	 */
+	for (match = rmap->head->match_list.head; match; match = match->next) {
+
+		if (!match->cmd || !match->cmd->str || !match->value)
+			continue;
+
+		ret = RMAP_NOMATCH;
+
+		afi = get_afi_from_match_rule(match->cmd->str);
+		if (afi == AFI_MAX)
+			return ret;
+
+		alist = access_list_lookup(afi, (char *)match->value);
+		if (!alist)
+			return ret;
+
+		/* If a match command refers to several objects in one
+		 * command either of them should match (i.e logical OR)
+		 */
+		FOREACH_ACCESS_LIST_FILTER(alist, alist_filter) {
```

Comment:
prefix-lists can be used in route-maps for exists-map/non-exists-map? Or only access-lists?

Reply:
Sorry I have added only access-list for now, will add prefix-list as well.